File: scripts/motor_config.py
#!/usr/bin/env python3
"""
motor_config.py - Shared motor configuration, loaded from scripts/config.json.

Every motor-driving script (controller.py, slider_controller.py, and any
future one) should call load_motor_config() instead of hardcoding its own
servo channel numbers, angle ranges, resting angles, or invert flags - so
config.json stays the single source of truth for the arm's physical limits.
"""

import json
from pathlib import Path

CONFIG_PATH = Path(__file__).resolve().parent / "config.json"

# Fallback per-motor settings, used for any motor missing from config.json
# (or if the file itself is missing/unreadable).
DEFAULT_MOTORS = {
    1: {"channel": 0, "min": 0, "max": 270, "rest": 135, "invert": False},
    2: {"channel": 1, "min": 0, "max": 270, "rest": 135, "invert": False},
    3: {"channel": 2, "min": 0, "max": 270, "rest": 135, "invert": False},
    4: {"channel": 3, "min": 0, "max": 270, "rest": 135, "invert": False},
    5: {"channel": 4, "min": 0, "max": 270, "rest": 135, "invert": False},
    6: {"channel": 5, "min": 0, "max": 270, "rest": 135, "invert": False},
}


def load_motor_config():
    """Return {motor_number: {"channel", "min", "max", "rest", "invert"}},
    merging config.json's "motors" section over DEFAULT_MOTORS."""
    motors = {n: dict(settings) for n, settings in DEFAULT_MOTORS.items()}

    if not CONFIG_PATH.exists():
        return motors

    try:
        with open(CONFIG_PATH, "r", encoding="utf-8") as f:
            data = json.load(f)
    except (OSError, json.JSONDecodeError) as e:
        print(f"Warning: could not read {CONFIG_PATH.name} ({e}), using default motor config.")
        return motors

    for key, overrides in data.get("motors", {}).items():
        try:
            n = int(key)
        except ValueError:
            continue
        motors.setdefault(n, dict(DEFAULT_MOTORS.get(n, DEFAULT_MOTORS[1])))
        motors[n].update(overrides)

    return motors

File: scripts/julia_controller/motor_config.jl
# motor_config.jl - Shared motor configuration, loaded from config.json at
# the repo root. Julia counterpart of scripts/motor_config.py - every
# motor-driving script here should call load_motor_config() instead of
# hardcoding servo channels/ranges, so config.json stays the single source
# of truth (shared with controller.py and the C++ controller).

import JSON3

const CONFIG_PATH = normpath(joinpath(@__DIR__, "..", "..", "config.json"))

const DEFAULT_MOTORS = Dict(
    1 => Dict("channel" => 0, "min" => 0, "max" => 270, "rest" => 135, "invert" => false, "kinematicSign" => 1),
    2 => Dict("channel" => 1, "min" => 0, "max" => 270, "rest" => 135, "invert" => false, "kinematicSign" => 1),
    3 => Dict("channel" => 2, "min" => 0, "max" => 270, "rest" => 135, "invert" => false, "kinematicSign" => 1),
    4 => Dict("channel" => 3, "min" => 0, "max" => 270, "rest" => 135, "invert" => false, "kinematicSign" => 1),
    5 => Dict("channel" => 4, "min" => 0, "max" => 270, "rest" => 135, "invert" => false, "kinematicSign" => 1),
    6 => Dict("channel" => 5, "min" => 0, "max" => 270, "rest" => 135, "invert" => false, "kinematicSign" => 1),
)

const DEFAULT_GEOMETRY = Dict(
    "units" => "mm",
    "baseHeight" => 0.0,
    "upperArmLength" => 0.0,
    "forearmLength" => 0.0,
    "wristLength" => 0.0,
)

function _read_config()
    isfile(CONFIG_PATH) || return nothing
    try
        return JSON3.read(read(CONFIG_PATH, String), Dict{String,Any})
    catch e
        println("Warning: could not read $(basename(CONFIG_PATH)) ($e), using defaults.")
        return nothing
    end
end

"""
    load_motor_config() -> Dict{Int,Dict{String,Any}}

Returns motor => {"channel","min","max","rest","invert","kinematicSign"},
merging config.json's "motors" section over DEFAULT_MOTORS.
"""
function load_motor_config()
    motors = Dict(n => copy(settings) for (n, settings) in DEFAULT_MOTORS)

    data = _read_config()
    data === nothing && return motors
    overrides_by_motor = get(data, "motors", Dict{String,Any}())

    for (key, overrides) in overrides_by_motor
        n = tryparse(Int, key)
        n === nothing && continue
        haskey(motors, n) || (motors[n] = copy(get(DEFAULT_MOTORS, n, DEFAULT_MOTORS[1])))
        for (k, v) in overrides
            motors[n][k] = v
        end
    end

    return motors
end

"""
    load_geometry() -> Dict{String,Any}

Returns the arm's link-length geometry (mm), merging config.json's
"geometry" section over DEFAULT_GEOMETRY.
"""
function load_geometry()
    geometry = copy(DEFAULT_GEOMETRY)

    data = _read_config()
    data === nothing && return geometry
    for (k, v) in get(data, "geometry", Dict{String,Any}())
        geometry[k] = v
    end

    return geometry
end

geometry_ready(geometry) = get(geometry, "upperArmLength", 0) != 0 && get(geometry, "forearmLength", 0) != 0
